File: Main2_73.java
import java.util.*;

public class Main2_73 {
    static int[] left;
    static int[] right;
    static List<Integer> preorderList = new ArrayList<>();
    static List<Integer> inorderList = new ArrayList<>();
    static List<Integer> postorderList = new ArrayList<>();

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();
        left = new int[n];
        right = new int[n];
        int[] parent = new int[n];

        for (int i = 0; i < n; i++) {
            left[i] = right[i] = parent[i] = -1;
        }

        for (int i = 0; i < n; i++) {
            int id = sc.nextInt();
            int l = sc.nextInt();
            int r = sc.nextInt();

            left[id] = l;
            right[id] = r;

            if (l != -1)
                parent[l] = id;
            if (r != -1)
                parent[r] = id;
        }

        int root = -1;
        for (int i = 0; i < n; i++) {
            if (parent[i] == -1) {
                root = i;
                break;
            }
        }

        preorder(root);
        inorder(root);
        postorder(root);

        System.out.println("Preorder");
        printList(preorderList);
        System.out.println("Inorder");
        printList(inorderList);
        System.out.println("Postorder");
        printList(postorderList);

        sc.close();
    }

    static void preorder(int u) {
        if (u == -1)
            return;
        preorderList.add(u);
        preorder(left[u]);
        preorder(right[u]);
    }

    static void inorder(int u) {
        if (u == -1)
            return;
        inorder(left[u]);
        inorderList.add(u);
        inorder(right[u]);
    }

    static void postorder(int u) {
        if (u == -1)
            return;
        postorder(left[u]);
        postorder(right[u]);
        postorderList.add(u);
    }

    static void printList(List<Integer> list) {
        for (int v : list) {
            System.out.print(" " + v);
        }
        System.out.println();
    }
}

File: Main2_73.py
import sys
sys.setrecursionlimit(10**7)

def preorder(u):
    if u == -1:
        return
    preorder_list.append(u)
    preorder(left[u])
    preorder(right[u])


def inorder(u):
    if u == -1:
        return
    inorder(left[u])
    inorder_list.append(u)
    inorder(right[u])


def postorder(u):
    if u == -1:
        return
    postorder(left[u])
    postorder(right[u])
    postorder_list.append(u)


def main():
    input = sys.stdin.readline

    n = int(input())

    global left, right
    left = [-1] * n
    right = [-1] * n
    parent = [-1] * n

    for _ in range(n):
        node_id, l, r = map(int, input().split())
        left[node_id] = l
        right[node_id] = r

        if l != -1:
            parent[l] = node_id
        if r != -1:
            parent[r] = node_id

    # root を探す
    root = -1
    for i in range(n):
        if parent[i] == -1:
            root = i
            break

    global preorder_list, inorder_list, postorder_list
    preorder_list = []
    inorder_list = []
    postorder_list = []

    preorder(root)
    inorder(root)
    postorder(root)

    print("Preorder")
    print(" " + " ".join(map(str, preorder_list)))

    print("Inorder")
    print(" " + " ".join(map(str, inorder_list)))

    print("Postorder")
    print(" " + " ".join(map(str, postorder_list)))


if __name__ == "__main__":
    main()